py/objexcept: Add uncatchable SystemAbort exception.

laurensvalk opened the pull request:
This adds an exception type that forces MicroPython to exit even if the user MicroPython script catches all exceptions.

This exception can be raised by C code to force MicroPython to exit for safety reasons, such as very low battery or very high temperatures.

I did not add it to the builtins dictionary to avoid adding new non-Python features to the language so it's just for internal use. Should there be a feature guard to avoid increasing codesize?

This is my first time looking around in `vm.c` so I'm not quite sure this is the right way to do it. Feel free to consider this as a draft for alternative implementations.

-----

**Context**

This issue was briefly discussed in the most recent MicroPython meetup. Damien's comments are at [26:00 onwards](https://youtu.be/M4m2V6FFuA8?t=1558).

![image](https://user-images.githubusercontent.com/12326241/201668040-2accfb24-7cc8-40dd-a132-82f6ba7e2767.png)

laurensvalk committed:
py/objexcept: Add uncatchable SystemAbort exception.

This adds an exception type that forces MicroPython to exit even if the
user MicroPython script catches all exceptions.

This exception can be raised by C code to force MicroPython to exit for
safety reasons, such as very low battery or very high temperatures.

laurensvalk commented:
Thanks Jim.

> in the slide you presented it looked like you weren't in ISR context

Indeed, we currently only use it on the main thread. For full context, we'd use it like this:

```c
// callback for when stop button is pressed (force_stop=false),
// or when shutdown is requested by firmware (force_stop=true)
void pbsys_main_stop_program(bool force_stop) {

    static const mp_obj_tuple_t args = {
        .base = { .type = &mp_type_tuple },
        .len = 1,
        .items = { MP_ROM_QSTR(MP_QSTR_stop_space_button_space_pressed) },
    };
    static mp_obj_exception_t system_exit;

    // Schedule SystemExit exception.
    system_exit.base.type = force_stop ? &mp_type_SystemAbort : &mp_type_SystemExit;
    system_exit.traceback_alloc = 0;
    system_exit.traceback_len = 0;
    system_exit.traceback_data = NULL;
    system_exit.args = (mp_obj_tuple_t *)&args;
    MP_STATE_MAIN_THREAD(mp_pending_exception) = MP_OBJ_FROM_PTR(&system_exit);
    #if MICROPY_ENABLE_SCHEDULER
    if (MP_STATE_VM(sched_state) == MP_SCHED_IDLE) {
        MP_STATE_VM(sched_state) = MP_SCHED_PENDING;
    }
    #endif
}
```
This code is mostly in use already, but in today's release we only have `mp_type_SystemExit`. The addition enabled by this PR is the `force_stop` that triggers `mp_type_SystemAbort`. (https://github.com/pybricks/pybricks-micropython/pull/117)

dpgeorge commented:
This looks very clean!  I thought it would be more difficult/intrusive.

If we accept this, it would need to be guarded by a config option, eg `MICROPY_ENABLE_SYSTEM_ABORT`.



> but I wondered whether we maybe need to consider cases where Python code is protected by a NLR scope that isn't managed by the VM.

In the case of Python IRQ handlers (eg scheduled functions) I don't think there is any problem.  Because already a `KeyboardInterrupt` is always correctly passed to the main thread, it's never handled by any scheduled handler.  This is because when a scheduled function runs it locks the scheduler, which also locks processing of asynchronous exceptions (which is what `KeyboardInterrupt` is, and also what `SystemAbort` could be).

laurensvalk commented:
> Yes that would be nicer.

Thanks for looking at this. I'll update the PR.

> But one issue with that there can only be one pending exception, so a KeyboardInterrupt can override a SystemAbort (and vice versa).

That is a good point, and it is something we're not currently handling above (thanks!).

But thinking about it now, it might even be beneficial for this use case. The firmware could call `mp_sched_system_abort` multiple times without problems, as opposed to `mp_raise_type(&mp_type_SystemAbort)` where more care is needed where it is called from and how many times.

laurensvalk committed:
py/scheduler: Schedule SystemExit or SystemAbort.

This implements mp_sched_system_exit_or_abort(), which can be used to
schedule a SystemExit or SystemAbort on the main thread. This can be
used by async sources to exit MicroPython if requested by a firmware
process. Use abort=true if the MicroPython script must not be able to
catch it, not even with a bare except.

The implementation is similar to mp_sched_keyboard_interrupt().

Signed-off-by: Laurens Valk <laurens@pybricks.com>

laurensvalk commented:
- Updated first commit to:
    - Add `MICROPY_ENABLE_SYSTEM_ABORT` config option as requested
    - Use `mp_obj_exception_match` instead of `mp_obj_is_type` (let me know if I should revert this)
- Added second commit to implement `mp_sched_system_exit_or_abort`

I tried to follow the pattern of `mp_sched_keyboard_interrupt` as much as possible; let me know if something else is preferred.

dpgeorge commented:
See alternative in #10241.

dpgeorge commented:
@jimmo and I had a long discussion about this today.  One thing that we think is not handled by the approach here is that it cannot abort the case where a user (accidentally) does a `while 1:pass` in a scheduled function.  That's because (1) scheduled exceptions are not processed in scheduled functions because the scheduler is locked at that time; (2) exceptions from scheduled functions are suppressed.

Also note that `mp_call_function_[12]_protected()` suppress any exceptions raised within them, so would suppress a `SystemAbort`.

If we are going to implement a "system abort" mechanism, I think it should be as robust as possible, ie able to abort (almost) all scenarios.

laurensvalk commented:
> If we are going to implement a "system abort" mechanism, I think it should be as robust as possible, ie able to abort (almost) all scenarios.

I completely agree. :+1: 

Don't worry about my PR. I suppose it's served its purpose by putting it on your radar :smile:

dpgeorge commented:
Closing in favour of #10241.
